test/spmd/test_xla_distributed_checkpoint.py

```diff
@@ -0,0 +1,165 @@
+                        save_planner=None,
+                        load_planner=None):
+    """
+    Checkpoint model_in using the provided save_planner and load into model_out
+    using the provided load_planner, and assert model_out equals model_in after
+    the load. If either planner is not specified, the DefaultPlanner is used.
+    """
+    tmpdir = tempfile.mkdtemp()
+
+    # Save an unsharded model using the provided save planner
+    dist_cp.save_state_dict(
+        state_dict=model_in.state_dict(),
+        storage_writer=dist_cp.FileSystemWriter(tmpdir),
+        planner=save_planner,
+        no_dist=True,  # Single-host checkpoint doesn't require a process group
+    )
+
+    # Load the checkpoint using the provided load planner
+    for p1, p2 in zip(model_in.parameters(), model_out.parameters()):
+      self.assertFalse(torch.allclose(p1, p2))
+    dist_cp.load_state_dict(
+        state_dict=model_out.state_dict(),
+        storage_reader=dist_cp.FileSystemReader(tmpdir),
+        planner=load_planner,
+        no_dist=True,  # Single-host checkpoint doesn't require a process group
+    )
+    for p1, p2 in zip(model_in.parameters(), model_out.parameters()):
+      self.assertTrue(torch.allclose(p1, p2))
+
+  def test_unsharded_to_sharded(self):
+    # Save an unsharded model using the DefaultSavePlanner and load into a
+    # sharded model using the SPMDLoadPlanner
+    model = self.SimpleLinear().to(xm.xla_device())
+    sharded_model = self._get_sharded_model()
+    self._save_and_restore(model, sharded_model, load_planner=SPMDLoadPlanner())
+
+
+class SPMDLoadPlannerTest(DistributedCheckpointTestBase):
+
+  def _get_load_planner(self, model):
+    # Create an SPMDLoadPlanner for the given model.
+    md = self._get_default_local_metadata()
+    planner = SPMDLoadPlanner()
+    planner.set_up_planner(model.state_dict(), md, True)
+    return planner
+
+  def test_state_dict_separation(self):
+    model = self._get_sharded_model()
+    planner = self._get_load_planner(model)
+    if self.n_devices > 1:
+      # The state_dict should be flattened and separated
+      self.assertCountEqual(planner.sharded_state_dict, ['fc1.weight'])
+      # `fc2.weight` should be in the unsharded_state_dict despite having
+      # an explicit mark_sharding call because it is replicated.
+      self.assertCountEqual(planner.unsharded_state_dict,
+                            ['fc1.bias', 'fc2.weight', 'fc2.bias'])
+    else:
+      # With a single device, no tensors are sharded.
+      self.assertCountEqual(
+          planner.unsharded_state_dict,
+          ['fc1.weight', 'fc1.bias', 'fc2.weight', 'fc2.bias'])
+
+  def test_local_load_plan(self):
+    model = self._get_sharded_model()
+    planner = self._get_load_planner(model)
+    plan = planner.create_local_plan()
+    parameter_count = len(list(model.parameters()))
+    if self.n_devices > 1:
+      # When the model is sharded across devices, fc1.weight will result in
```

Comment:
So you suggest the number of ReadItem will match the number of XLAShard for a particular XLAShardedTensor. Does this hold true for the following resharding example?
1. stored shards <= loading shard, i.e., the models is sharded in 2 ways but need to loaded into 4 ways.
2. stored shards > loading shard, i.e., the models is sharded in 4 ways but need to loaded into 2 ways.

I can see that holds true for case 1 but it's hard for me to imagine for case 2. Can you elaborate? Commented in your design as well.

Reply:
That's true - I'll clarify the comment that this comes from using the unsharded checkpoint metadata used when creating the Planner in `_get_load_planner`.

Reply:
If we are loading into a coarser mesh, e.g. loading a checkpoint taken on a (4, 4) mesh into a model sharded across (2, 2), each local shard in the (2, 2) mesh will require multiple ReadItems. On our side, we will still generate a single ChunkStorageMetadata for each shard, and the utility function `create_read_items_for_chunk_list` translates these into the ReadItems necessary by finding the overlap of `ChunkStorageMetadata` with the chunks in the storage metadata.

The algorithm for `create_read_items_for_chunk_list` just iterates across all chunks in the storage metadata, identifies overlap with the `ChunkStorageMetadata` describing our XLAShard, and generates a ReadItem if there is overlap.

Reply:
High level speaking, the number of ReadItems will then be determined by the max(number of shards in loading model, number of shards in saving model)?

Reply:
Perfect. Thanks, Jon!